presto-main/src/main/java/com/facebook/presto/util/PrestoIterators.java

```diff
@@ -0,0 +1,66 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.util;
+
+import com.google.common.collect.AbstractIterator;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.Iterator;
+
+public class PrestoIterators
+{
+    public static <T> Iterator<T> asCloseable(Iterator<T> iterator, Closeable resource)
+    {
+        return new CloseableIterator<>(iterator, resource);
+    }
+
+    public static class CloseableIterator<T>
+            extends AbstractIterator<T>
+            implements AutoCloseable
```

Comment:
It doesn't help to declare this `public` and `implements AutoCloseable`. This is later cast to `Iterator<T>` (e.g. `Iterator<Page>`) and users won't know there is `close()` method and that it `must` be called.

This could be solved by:
* making `CloseableIterator` an interface and using it everywhere down the call path; then `close()` should be explicit and not "when exhausted"
* making this class `private`, unimplementing `AutoCloseable`

In our particular situation I prefer the latter.